packages/captp/src/captp.js

```diff
@@ -0,0 +1,703 @@
+        });
+      });
+    },
+    // Remote is invoking a method or retrieving a property.
+    async CTP_CALL(obj) {
+      // questionId: Remote promise (for promise pipelining) this call is
+      //   to fulfill
+      // target: Slot id of the target to be invoked.  Checks against
+      //   answers first; otherwise goes through unserializer
+      const { questionID, target, trap } = obj;
+
+      const [prop, args] = unserialize(obj.method);
+      let val;
+      if (answers.has(target)) {
+        val = answers.get(target);
+      } else {
+        val = unserialize({
+          body: JSON.stringify({
+            [QCLASS]: 'slot',
+            index: 0,
+          }),
+          slots: [target],
+        });
+      }
+
+      /** @type {(isReject: boolean, value: any) => void} */
+      let sendReturn = (isReject, value) => {
+        send({
+          type: 'CTP_RETURN',
+          epoch,
+          answerID: questionID,
+          [isReject ? 'exception' : 'result']: serialize(harden(value)),
+        });
+      };
+      if (trap) {
+        try {
+          assert(
+            exportedTrapHandlers.has(val),
+            X`Refused Trap(${val}) because target was not registered with makeTrapHandler`,
+          );
+          assert.typeof(
+            trapHost,
+            'function',
+            X`CapTP cannot answer Trap(${val}) without a trapHost function`,
+          );
+
+          // We need to create a promise for the "isDone" iteration right now to
+          // prevent a race with the other side.
+          const resultPK = makePromiseKit();
+          trapIteratorResultP.set(questionID, resultPK.promise);
+
+          sendReturn = async (isReject, value) => {
+            const serialized = serialize(harden(value));
+            const ait = trapHost([isReject, serialized]);
+            if (!ait) {
+              // One-shot, no async iterator.
+              resultPK.resolve({ done: true });
+              return;
+            }
+
+            // We're ready for them to drive the iterator.
+            trapIterator.set(questionID, ait);
+            resultPK.resolve({ done: false });
+          };
+        } catch (e) {
+          sendReturn(true, e);
+          throw e;
+        }
+      }
+
+      // If `args` is supplied, we're applying a method or function... otherwise this is
+      // property access
+      let hp;
+      if (!args) {
+        hp = HandledPromise.get(val, prop);
+      } else if (prop === null) {
+        hp = HandledPromise.applyFunction(val, args);
+      } else {
+        hp = HandledPromise.applyMethod(val, prop, args);
+      }
+
+      // Answer with our handled promise
+      answers.set(questionID, hp);
+
+      // Set up promise resolver for this handled promise to send
+      // message to other vat when fulfilled/broken.
+      return hp
+        .then(res => sendReturn(false, res))
+        .catch(rej => sendReturn(true, rej))
+        .catch(rej => quietReject(rej, false));
```

Comment:
When I see `res` and `rej` I assume I'm looking at resolve and reject functions.
Did you really want the first catch to also catch errors from the then callback?

```suggestion
      return hp
        .then(
          fulfilment => sendReturn(false, fulfilment),
          reason => sendReturn(true, reason)
        ).catch(reason => quietReject(reason, false));
```

Reply:
Good point.  Every time I write such handlers, I have to rediscover the error semantics I should be aiming for.

Reply:
I simplified: our caller (the `dispatch` function) already handles rejections, so I just ensure this result isn't dropped on the floor.